examples_tests/22.RaytracedAO/raytraceCommon.glsl

```diff
@@ -252,14 +260,82 @@ vec3 load_normal_and_prefetch_textures(
 	return geomNormal;
 }
 
+vec3 nbl_glsl_unormSphericalToCartesian(in vec2 uv, out float sinTheta)
+{
+	vec3 dir;
+	nbl_glsl_sincos((uv.x-0.5)*2.f*nbl_glsl_PI,dir.y,dir.x);
+	nbl_glsl_sincos(uv.y*nbl_glsl_PI,sinTheta,dir.z);
+	dir.xy *= sinTheta;
+	return dir;
+}
+
+// return regularized pdf of sample
+float Envmap_regularized_deferred_pdf(in vec3 rayDirection)
+{
+	const ivec2 luminanceMapSize = textureSize(luminance[0], 0);
+	uint lastLuminanceMip = uint(log2(luminanceMapSize.x)); // TODO: later turn into push constant
+	const vec2 envmapUV = nbl_glsl_sampling_generateUVCoordFromDirection(rayDirection);
+
+	float sinTheta = length(rayDirection.zx);
+	float sumLum = texelFetch(luminance[lastLuminanceMip], ivec2(0), 0).r;
+	float lum = textureLod(luminance[0], envmapUV, 0).r;
+	float bigfactor = float(luminanceMapSize.x*luminanceMapSize.y)/sumLum;
+	return bigfactor*(lum/(sinTheta*2.0f*nbl_glsl_PI*nbl_glsl_PI));
+}
```

Comment:
you can yeet this to a header (`nbl_glsl_sampling_envmap_HierarchicalWarp_`) if you make it take a `sampler2D` argument

Reply:
Also an override which takes an extra `in ivec2 luminanceBaseMipLevelSize, in int luminanceHighestMipLevel` would be great, as to not call `textureSize` or `log2` needlessly

Reply:
Done